File: MaiChat/desktop/tests/ChatStateTest.cpp
#include <QtTest/QtTest>

#include "model/ChatState.h"

class ChatStateTest : public QObject {
    Q_OBJECT

private slots:
    void queuesOutgoingText();
    void removesAicliProtocolPrefixFromIncomingText();
    void removesAicliProtocolPrefixFromRestoredIncomingText();
    void receivesIncomingImage();
    void receivesIncomingFile();
    void updatesMessageStatus();
    void returnsPeerMessagesChronologically();
    void reordersSameSecondMessagesAfterTimeCorrection();
    void removesContactAndMessages();
    void adoptsRemoteMessageIdAndDropsDuplicateTemp();
    void countsUnreadOnlyForBackgroundPeers();
    void selectPeerClearsUnread();
    void restoreDoesNotAffectUnread();
    void removingContactDropsItsUnread();
};

void ChatStateTest::queuesOutgoingText() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "ios-user"});
    state.selectPeer("ios-user");

    const RemoteIMMessage message = state.queueOutgoingText("ping");

    QCOMPARE(message.fromUserId, QString("desktop-user"));
    QCOMPARE(message.toUserId, QString("ios-user"));
    QCOMPARE(message.text, QString("ping"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Outgoing);
    QCOMPARE(message.status, RemoteIMMessageStatus::Pending);
    QCOMPARE(message.createdAtMillis % 1000, Q_INT64_C(0));
    QCOMPARE(state.messagesWith("ios-user").size(), 1);
}

void ChatStateTest::removesAicliProtocolPrefixFromIncomingText() {
    ChatState state("desktop-user");
    const QString hiddenPrefix = QStringLiteral("\u2063\u200B\u200C\u200D\u2063");

    const RemoteIMMessage hiddenMarked = state.receiveText(
        "electron-user",
        hiddenPrefix + QStringLiteral("## 结论\n\nMarkdown 正文")
    );
    const RemoteIMMessage legacyMarked = state.receiveText(
        "electron-user",
        QStringLiteral("【AICLI 输出】\n## 结果")
    );

    QCOMPARE(hiddenMarked.text, QStringLiteral("## 结论\n\nMarkdown 正文"));
    QCOMPARE(legacyMarked.text, QStringLiteral("## 结果"));
}

void ChatStateTest::removesAicliProtocolPrefixFromRestoredIncomingText() {
    ChatState state("desktop-user");
    const QString hiddenPrefix = QStringLiteral("\u2063\u200B\u200C\u200D\u2063");
    RemoteIMMessage message;
    message.fromUserId = QStringLiteral("electron-user");
    message.toUserId = QStringLiteral("desktop-user");
    message.text = hiddenPrefix + QStringLiteral("# Win/Mac 每周 Crash 详细报表");
    message.direction = RemoteIMMessageDirection::Incoming;

    state.appendMessageForRestore(message);

    QCOMPARE(state.messagesWith(QStringLiteral("electron-user")).first().text,
             QStringLiteral("# Win/Mac 每周 Crash 详细报表"));
}

void ChatStateTest::receivesIncomingImage() {
    ChatState state("desktop-user");

    const RemoteIMMessage message = state.receiveImage("ios-user", "/tmp/a.jpg", 640, 480, 1024);

    QCOMPARE(message.fromUserId, QString("ios-user"));
    QCOMPARE(message.toUserId, QString("desktop-user"));
    QCOMPARE(message.text, QString("[图片消息] a.jpg"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Incoming);
    QCOMPARE(message.status, RemoteIMMessageStatus::Received);
    QVERIFY(message.hasImage);
    QCOMPARE(state.contacts().size(), 1);
    QCOMPARE(state.selectedPeerId(), QString("ios-user"));
}

void ChatStateTest::receivesIncomingFile() {
    ChatState state("desktop-user");

    const RemoteIMMessage message = state.receiveFile(
        "ios-user",
        "/tmp/report.md",
        "report.md",
        "text/markdown",
        4096
    );

    QCOMPARE(message.fromUserId, QString("ios-user"));
    QCOMPARE(message.toUserId, QString("desktop-user"));
    QCOMPARE(message.text, QString("[文件消息] report.md"));
    QCOMPARE(message.direction, RemoteIMMessageDirection::Incoming);
    QCOMPARE(message.status, RemoteIMMessageStatus::Received);
    QVERIFY(message.hasFile);
    QCOMPARE(message.file.localPath, QString("/tmp/report.md"));
    QCOMPARE(message.file.fileName, QString("report.md"));
    QCOMPARE(message.file.mimeType, QString("text/markdown"));
    QCOMPARE(message.file.sizeBytes, qint64(4096));
}

void ChatStateTest::updatesMessageStatus() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "ios-user"});
    state.selectPeer("ios-user");
    const RemoteIMMessage message = state.queueOutgoingText("ping");

    QVERIFY(state.updateMessageStatus(message.id, RemoteIMMessageStatus::Sent));

    QCOMPARE(state.messagesWith("ios-user").first().status, RemoteIMMessageStatus::Sent);
}

void ChatStateTest::returnsPeerMessagesChronologically() {
    ChatState state("desktop-user");

    RemoteIMMessage newest;
    newest.id = QStringLiteral("newest");
    newest.fromUserId = QStringLiteral("ios-user");
    newest.toUserId = QStringLiteral("desktop-user");
    newest.text = QStringLiteral("30s");
    newest.createdAtMillis = 30'000;

    RemoteIMMessage oldest = newest;
    oldest.id = QStringLiteral("oldest");
    oldest.text = QStringLiteral("10s");
    oldest.createdAtMillis = 10'000;

    RemoteIMMessage middle = newest;
    middle.id = QStringLiteral("middle");
    middle.text = QStringLiteral("20s");
    middle.createdAtMillis = 20'000;

    state.appendMessageForRestore(newest);
    state.appendMessageForRestore(oldest);
    state.appendMessageForRestore(middle);

    const QList<RemoteIMMessage> messages = state.messagesWith(QStringLiteral("ios-user"));

    QCOMPARE(messages.size(), 3);
    QCOMPARE(messages.at(0).text, QStringLiteral("10s"));
    QCOMPARE(messages.at(1).text, QStringLiteral("20s"));
    QCOMPARE(messages.at(2).text, QStringLiteral("30s"));
}

void ChatStateTest::reordersSameSecondMessagesAfterTimeCorrection() {
    ChatState state("desktop-user");

    RemoteIMMessage request;
    request.id = QStringLiteral("z-request");
    request.fromUserId = QStringLiteral("ios-user");
    request.toUserId = QStringLiteral("desktop-user");
    request.text = QStringLiteral("检查编译流程");
    request.direction = RemoteIMMessageDirection::Incoming;
    request.createdAtMillis = 10'482;

    RemoteIMMessage ack = request;
    ack.id = QStringLiteral("a-ack");
    ack.fromUserId = QStringLiteral("desktop-user");
    ack.toUserId = QStringLiteral("ios-user");
    ack.text = QStringLiteral("已发送给当前 AICLI，开始处理。");
    ack.direction = RemoteIMMessageDirection::Outgoing;
    ack.createdAtMillis = 10'000;

    // 本机乐观消息带毫秒，随后收到的 SDK 消息只有秒级时间：确认信息回写前
    // 会错误地把后发回执排在用户消息前面。
    state.appendMessageForRestore(request);
    state.appendMessageForRestore(ack);
    QCOMPARE(state.messagesWith(QStringLiteral("ios-user")).first().id, QStringLiteral("a-ack"));

    QVERIFY(state.updateMessageTime(QStringLiteral("z-request"), 10'000));

    const QList<RemoteIMMessage> messages = state.messagesWith(QStringLiteral("ios-user"));
    QCOMPARE(messages.size(), 2);
    QCOMPARE(messages.at(0).id, QStringLiteral("z-request"));
    QCOMPARE(messages.at(1).id, QStringLiteral("a-ack"));
}

void ChatStateTest::removesContactAndMessages() {
    ChatState state("desktop-user");
    state.upsertContact(RemoteIMContact{"ios-user", "iPhone"});
    state.upsertContact(RemoteIMContact{"other-user", "Other"});
    state.selectPeer("ios-user");
    state.receiveText("ios-user", "hello");
    state.receiveText("other-user", "keep");

    state.removeContactAndMessages(" ios-user ");

    QCOMPARE(state.contacts().size(), 1);
    QCOMPARE(state.contacts().first().userId, QString("other-user"));
    QCOMPARE(state.messagesWith("ios-user").size(), 0);
    QCOMPARE(state.messagesWith("other-user").size(), 1);
    QCOMPARE(state.selectedPeerId(), QString("other-user"));
}

void ChatStateTest::adoptsRemoteMessageIdAndDropsDuplicateTemp() {
    ChatState state("me");
    state.upsertContact(RemoteIMContact{"peer", "peer"});
    state.selectPeer("peer");

    // 常规采纳：临时 UUID 换成 SDK 稳定 id，状态更新按新 id 生效。
    const RemoteIMMessage queued = state.queueOutgoingText("hello");
    QVERIFY(state.adoptMessageId(queued.id, "sdk-1#0"));
    QVERIFY(state.updateMessageStatus("sdk-1#0", RemoteIMMessageStatus::Sent));
    QCOMPARE(state.messagesWith("peer").first().id, QStringLiteral("sdk-1#0"));

    // 稳定 id 已存在（漫游先到）：采纳失败并移除临时重复项。
    RemoteIMMessage roamed;
    roamed.id = "sdk-2#0";
    roamed.fromUserId = "me";
    roamed.toUserId = "peer";
    roamed.direction = RemoteIMMessageDirection::Outgoing;
    roamed.status = RemoteIMMessageStatus::Sent;
    roamed.text = "dup";
    state.appendMessageForRestore(roamed);
    const RemoteIMMessage temp = state.queueOutgoingText("dup");
    QVERIFY(!state.adoptMessageId(temp.id, "sdk-2#0"));
    int dupCount = 0;
    for (const RemoteIMMessage& message : state.messagesWith("peer")) {
        if (message.text == QStringLiteral("dup")) ++dupCount;
    }
    QCOMPARE(dupCount, 1);
}

void ChatStateTest::countsUnreadOnlyForBackgroundPeers() {
    ChatState state("desktop-user");

    // 首条消息自动选中该会话，视为已读：不计红点。
    state.receiveText("peer-a", "hello");
    QCOMPARE(state.selectedPeerId(), QString("peer-a"));
    QCOMPARE(state.unreadCount("peer-a"), 0);

    // 非选中会话的实时消息累计未读（文本/图片/文件/语音同规则）。
    state.receiveText("peer-b", "one");
    state.receiveImage("peer-b", "C:/tmp/pic.png", 10, 10, 100);
    state.receiveFile("peer-b", "C:/tmp/report.pdf", "report.pdf", "application/pdf", 200);
    state.receiveVoice("peer-b", "C:/tmp/voice.mp3", 3);
    QCOMPARE(state.unreadCount("peer-b"), 4);
    QCOMPARE(state.unreadCount("peer-a"), 0);

    // 当前选中会话继续收消息不计未读（消息就在屏幕上）。
    state.receiveText("peer-a", "again");
    QCOMPARE(state.unreadCount("peer-a"), 0);
}

void ChatStateTest::selectPeerClearsUnread() {
    ChatState state("desktop-user");
    state.receiveText("peer-a", "hello");
    state.receiveText("peer-b", "hi");
    state.receiveText("peer-b", "there");
    QCOMPARE(state.unreadCount("peer-b"), 2);

    state.selectPeer("peer-b");

    QCOMPARE(state.unreadCount("peer-b"), 0);
    // 切走后新消息重新计数。
    state.selectPeer("peer-a");
    state.receiveText("peer-b", "back");
    QCOMPARE(state.unreadCount("peer-b"), 1);
}

void ChatStateTest::restoreDoesNotAffectUnread() {
    ChatState state("desktop-user");
    state.selectPeer("peer-a");

    // 本地库加载/SDK 漫游补充是历史消息，不产生红点。
    RemoteIMMessage history;
    history.fromUserId = QStringLiteral("peer-b");
    history.toUserId = QStringLiteral("desktop-user");
    history.text = QStringLiteral("yesterday");
    history.direction = RemoteIMMessageDirection::Incoming;
    state.appendMessageForRestore(history);

    QCOMPARE(state.unreadCount("peer-b"), 0);
}

void ChatStateTest::removingContactDropsItsUnread() {
    ChatState state("desktop-user");
    state.receiveText("peer-a", "hello");
    state.receiveText("peer-b", "hi");
    QCOMPARE(state.unreadCount("peer-b"), 1);

    state.removeContactAndMessages("peer-b");
    QCOMPARE(state.unreadCount("peer-b"), 0);

    // 再次加回联系人不携带历史红点。
    state.receiveText("peer-b", "fresh");
    QCOMPARE(state.unreadCount("peer-b"), 1);
    state.selectPeer("peer-b");
    QCOMPARE(state.unreadCount("peer-b"), 0);
}

QTEST_MAIN(ChatStateTest)
#include "ChatStateTest.moc"

File: MaiChat/android/MaiChat/app/src/test/java/com/kongshang/maichat/ChatStateTest.java
package com.kongshang.maichat;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import java.util.List;

public class ChatStateTest {
    @Test
    public void contactGroupsRenameDeleteAndPreserveProfileAssignments() {
        ChatState state = new ChatState("android-user");
        assertTrue(state.addContactGroup(" 同事 "));
        assertTrue(state.addContactGroup("未分组"));
        assertFalse(state.addContactGroup("同事"));
        assertFalse(state.addContactGroup("  "));

        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        assertTrue(state.setContactGroup("mac-office", "同事"));
        state.upsertContact(new RemoteIMContact(
            "mac-office", "办公室电脑", "https://example.com/avatar.png"
        ));
        assertEquals("同事", state.contacts().get(0).groupName());
        assertEquals("办公室电脑", state.contacts().get(0).displayName());

        assertTrue(state.renameContactGroup("同事", "工作"));
        assertEquals("工作", state.contacts().get(0).groupName());
        assertFalse(state.renameContactGroup("工作", "未分组"));
        assertTrue(state.removeContactGroup("工作"));
        assertEquals("", state.contacts().get(0).groupName());
        assertEquals(List.of("未分组"), state.contactGroups());
    }

    @Test
    public void unknownContactGroupSelfHealsToUngrouped() {
        ChatState state = new ChatState("android-user");
        state.setContactGroups(List.of("同事"));
        state.upsertContact(new RemoteIMContact("mac-office", "Mac", "", "幽灵组"));

        assertEquals("", state.contacts().get(0).groupName());
        assertTrue(state.setContactGroup("mac-office", "幽灵组"));
        assertEquals("", state.contacts().get(0).groupName());
    }

    @Test
    public void upsertsContactAndQueuesOutgoingText() {
        ChatState state = new ChatState("android-user");

        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");
        RemoteIMMessage message = state.queueOutgoingText("检查构建");

        assertEquals("mac-office", state.selectedPeerId());
        assertEquals(1, state.contacts().size());
        assertEquals("检查构建", message.text());
        assertEquals(RemoteIMMessage.Direction.OUTGOING, message.direction());
        assertEquals(RemoteIMMessage.Status.PENDING, message.status());
        assertEquals(List.of(message), state.messagesWith("mac-office"));
    }

    @Test
    public void recordsIncomingTextAndAutoAddsContact() {
        ChatState state = new ChatState("android-user");

        RemoteIMMessage message = state.receiveText("处理完成", "mac-office");

        assertEquals("处理完成", message.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, message.direction());
        assertEquals(RemoteIMMessage.Status.RECEIVED, message.status());
        assertEquals("mac-office", state.contacts().get(0).userId());
    }

    @Test
    public void queuesOutgoingImageWithAttachment() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");

        RemoteIMMessage message = state.queueOutgoingImage("/tmp/photo.png", 640, 480, 4096);

        assertEquals("[图片消息] photo.png", message.text());
        assertNotNull(message.imageAttachment());
        assertEquals("/tmp/photo.png", message.imageAttachment().localPath());
        assertEquals(640, message.imageAttachment().width());
        assertEquals(480, message.imageAttachment().height());
        assertEquals(4096, message.imageAttachment().sizeBytes());
    }

    @Test
    public void queuesAndReceivesVoiceMessages() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");

        RemoteIMMessage outgoing = state.queueOutgoingVoice("/tmp/out.m4a", 3);
        RemoteIMMessage incoming = state.receiveVoice("/tmp/in.m4a", 4, "mac-office");

        assertEquals("[语音消息 3s]", outgoing.text());
        assertEquals(RemoteIMMessage.Direction.OUTGOING, outgoing.direction());
        assertNotNull(outgoing.voiceAttachment());
        assertEquals("[语音消息 4s]", incoming.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, incoming.direction());
        assertNotNull(incoming.voiceAttachment());
    }

    @Test
    public void receivesMarkdownFileWithAttachment() {
        ChatState state = new ChatState("android-user");

        RemoteIMMessage message = state.receiveFile(
            "/tmp/remote-im/report.md",
            "mac-office",
            "report.md",
            "text/markdown",
            4096
        );

        assertEquals("[文件消息] report.md", message.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, message.direction());
        assertEquals(RemoteIMMessage.Status.RECEIVED, message.status());
        assertNotNull(message.fileAttachment());
        assertEquals("/tmp/remote-im/report.md", message.fileAttachment().localPath());
        assertEquals("report.md", message.fileAttachment().fileName());
        assertEquals("text/markdown", message.fileAttachment().mimeType());
        assertEquals(4096, message.fileAttachment().sizeBytes());
    }

    @Test
    public void messageStatusCanBeUpdated() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));
        state.selectPeer("mac-office");
        RemoteIMMessage message = state.queueOutgoingText("ping");

        assertTrue(state.updateMessageStatus(message.id(), RemoteIMMessage.Status.SENT));

        assertEquals(RemoteIMMessage.Status.SENT, state.messagesWith("mac-office").get(0).status());
        assertFalse(state.updateMessageStatus("missing", RemoteIMMessage.Status.SENT));
    }

    @Test
    public void deduplicatesIncomingMessagesByRemoteId() {
        ChatState state = new ChatState("android-user");

        RemoteIMMessage first = state.receiveText(
            "完成",
            "mac-office",
            "remote-1",
            100L,
            RemoteIMOrigin.MACHINE
        );
        RemoteIMMessage duplicate = state.receiveText(
            "重复",
            "mac-office",
            "remote-1",
            200L,
            RemoteIMOrigin.MACHINE
        );

        assertEquals(first, duplicate);
        assertEquals(1, state.messagesWith("mac-office").size());
        assertEquals(RemoteIMOrigin.MACHINE, first.origin());
    }

    @Test
    public void deletingAContactAlsoRemovesOnlyThatConversation() {
        ChatState state = new ChatState("android-user");
        state.receiveText("A", "mac-office");
        state.receiveText("B", "house-office");

        assertTrue(state.removeContact("mac-office"));

        assertTrue(state.messagesWith("mac-office").isEmpty());
        assertEquals(1, state.messagesWith("house-office").size());
    }

    @Test
    public void returnsPeerMessagesChronologically() {
        ChatState state = new ChatState("android-user");
        RemoteIMMessage newest = new RemoteIMMessage(
            "newest",
            "android-user",
            "mac-office",
            "最新消息",
            RemoteIMMessage.Direction.OUTGOING,
            RemoteIMMessage.Status.SENT,
            300L,
            null,
            null
        );
        RemoteIMMessage oldest = new RemoteIMMessage(
            "oldest",
            "mac-office",
            "android-user",
            "最早消息",
            RemoteIMMessage.Direction.INCOMING,
            RemoteIMMessage.Status.RECEIVED,
            100L,
            null,
            null
        );
        RemoteIMMessage middle = new RemoteIMMessage(
            "middle",
            "mac-office",
            "android-user",
            "中间消息",
            RemoteIMMessage.Direction.INCOMING,
            RemoteIMMessage.Status.RECEIVED,
            200L,
            null,
            null
        );
        state.addRestoredMessage(newest);
        state.addRestoredMessage(oldest);
        state.addRestoredMessage(middle);

        assertEquals(List.of(oldest, middle, newest), state.messagesWith("mac-office"));
    }

    @Test
    public void receivesVideoAndDedupesByRemoteId() {
        ChatState state = new ChatState("android-user");
        state.upsertContact(new RemoteIMContact("mac-office", "Mac Office"));

        RemoteIMMessage message = state.receiveVideo(
            "/tmp/clip.mp4",
            "/tmp/clip-cover.jpg",
            12,
            1080,
            1920,
            4_194_304L,
            "mac-office",
            "sdk-video-1",
            1_700_000_000_000L,
            RemoteIMOrigin.HUMAN
        );

        assertEquals("[视频消息 12s]", message.text());
        assertEquals(RemoteIMMessage.Direction.INCOMING, message.direction());
        assertNotNull(message.videoAttachment());
        assertEquals("/tmp/clip-cover.jpg", message.videoAttachment().coverPath());

        // 同一条 remoteId 再来一次不能变成两条：封面与视频是两次下载，
        // 上层对同一条消息可能不止投递一次。
        RemoteIMMessage again = state.receiveVideo(
            "/tmp/clip.mp4",
            "/tmp/clip-cover.jpg",
            12,
            1080,
            1920,
            4_194_304L,
            "mac-office",
            "sdk-video-1",
            1_700_000_000_000L,
            RemoteIMOrigin.HUMAN
        );

        assertEquals(message.id(), again.id());
        assertEquals(1, state.messagesWith("mac-office").size());
    }
}
